File: mcdowell/python/ch1-06.py
import pprint

def rotate(matrix):

    width = len(matrix)

    # layer = vertical column
    for layer in range(width // 2):
        first = layer
        last = width - 1 - layer
        for i in range(first, last):
            offset = i - first

            # top
            matrix[first][i]

            # right
            matrix[i][last]

            #bottom
            matrix[last][last - offset]

            #left
            matrix[last - offset][first]



            top = matrix[first][i]

            # top <- left
            matrix[first][i] = matrix[last - offset][first]

            # left <- bottom
            matrix[last - offset][first] = matrix[last][last - offset]

            # bottom <- right
            matrix[last][last - offset] = matrix[i][last]

            # right <- top
            matrix[i][last] = top



# reads as ans[y][x]
ans = [
    [1,2,3,4,5], # 1st row
    [1,2,3,4,5], # 2nd row
    [1,2,3,4,5],
    [1,2,3,4,5],
    [1,2,3,4,5],
]
rotate(ans)
pprint.pprint(ans)

File: mcdowell/js/ch1-06.js


function rotateMatrix2D(m) {

    var width = m.length;

    for (var layer = 0; layer < Math.floor(width/2); layer ++ ) {
        var first = layer;
        var last = width - 1 - layer;

        for ( var i = first; i < last; i ++) {

            var offset = i - first;

            // m[ y ][ x ]

            // top = m[first][i]
            // right = m[i][last]
            // bottom = m[last][last - offset]
            // left = m[last - offset][first]

            var top = m[first][i];

            // top <- left
            m[first][i] = m[last - offset][first];

            // left <- bottom
            m[last - offset][first] = m[last][last - offset];

            // bottom <- right
            m[last][last - offset] = m[i][last];

            // right <- top
            m[i][last] = top;


        }
    }
}


// ans[y][x]
var ans = [
    [1,2,3,4,5], // 1st row
    [1,2,3,4,5],
    [1,2,3,4,5],
    [1,2,3,4,5],
    [1,2,3,4,5],
];

rotateMatrix2D(ans);

console.log(ans)
